### DISCUSSION TITLE
GPIO Pin assignments for radio

### DISCUSSION THREAD

**Discussion by acutetech** [azerimaker/Penguino-STM32WL-LoRa-E5#8] [General]
Hi Orkhan
I have asked Seeed the following question but not received a reply. Are you able to help?

In order to develop with this module with ST's ST32CUBE IDE development environment further details are essential: which GPIO pins are used to control the TCXO and RF switch, and what do they do? Your radio_board_if.h and .c files will provide this information - can I have these please?

**Comment by azerimaker** [comment]
Hi, @acutetech. Thanks for contacting. 
To be honest I still haven't had a chance to test the LoRaWAN code, only managed to flash the RGB LED for now. I'm guessing the pinout is the same for [NUCLEO-WL55JC](https://www.st.com/en/evaluation-tools/nucleo-wl55jc.html). You can download the design files and have a look at the pin assignments for TCXO and the RF switch control lines. 

Hope it helps, 

Cheers

**Reply by acutetech** [reply]
Thanks Orkhan
I don't think we can assume the LoRa-E5 has the same pin assignmant as the NUCLEO-WL55JC. Certainly the published use of pins differs: pins used for buttons and LEDs are not available on the LoRa-E5. The NUCLEO-WL55JC uses 3 pins to control RF switches, including switching between high and low power outputs, plus one to control the TCXO. IMHO the LoRa-E5 is likely to have a simpler scheme, such as one bit that toggles between transmit and receive.

Do you have contacts at Seeed who can tell you? Or ask them to publish the schematic? They have not answered me (even my question is not visible on their forum). This is essential information to use the module. My fall-back position is to remove the components and probe the PCB, but it should not have to come to that...

Regards - Charles

**Reply by azerimaker** [reply]
This is what I got from my contact at Seeed. I hope it helps. 

![WhatsApp Image 2021-03-04 at 14 26 50](https://user-images.githubusercontent.com/2662257/109976698-5c797a80-7cfc-11eb-99c7-dc8e2c677888.jpeg)

Cheers, 
Orkhan

**Comment by acutetech** [comment]
Thanks for that. Two comments and two questions: 

(a) I find it curious that PB5 is available externally but PA4 is not. 
(b) It is a shame they used PB5 to control the RF switch since that prevents the use of SPI1 (PB5 was the only SPI1_MOSI pin made available externally).

The two questions: 

(q1) Given the above, and that their data sheet lists PB5 as "MCU GPIO" with no restriction on its use, I wonder if the pin assignments you sent me are correct?
(q2) There is still insufficient information here to program the module: what state do we set the pins when neither transmitting nor receiving?

It would be much more helpful if Seeed would release either the circuit diagram or the radio_board_if.h and .c files. Perhaps you can pass on this request.

**Comment by acutetech** [comment]
Hi Orkhan

I just thought you might be interested:

A few minutes ago I transferred my first LoRaWan messages to The Thing Network, using the LoRa-E5 on my own PCB. I am using ST's "LoRaWan_End_Node" application code, modified for my hardware, and making adjustments for the LoRa-E5 module. These included the radio pin assignment inforation you sent me. There were many problems to solve on the way, and much testing remaining, but I thought it is worth reporting here that it can be done!

I had to make guesses on a number of questions, and I am upset that Seeed refuse to share their radio_board_if.h & .c

Have you managed to communicate on your own board? If so, can we exchange radio_board_if.h & .c files?

**Comment by azerimaker** [comment]
Hi, @acutetech (Charles). 
This is a great news. Thanks for the update. 
I still haven't tried making it work, maybe next week over the Easter break. If you can, please do a PR or fork of this repo and add your LoRaWAN example to it, so I can try to verify it myself. 
Another option would be to try to port the [Generic-Node-SE repo](https://github.com/TheThingsIndustries/generic-node-se/). I'll follow up with more updates.

**Comment by acutetech** [comment]
Hi Orkhan

Rather than your suggestions I will attach some files here and you can see what you can do with them. My notes to myself follow:

So I could accomodate different boards, I created different board_resources_xxx.h files for each board, 
then I can choose which to include with macros:

board_resources.h now specifies one of:

- 	board_resources_nucleo.h	(original settings provided by ST - untested)	
- 	board_resources_grovee5.h	(For the LoRa-E5 on Grove-Lora_e5 board - partly tested)
- 	board_resources_ww120.h		(For the WW120 - my own hardware)

	board_resources.c rearranges the sequence of LEDs and buttons, allows for LED polarity, includes some debug messages
		
 The switches are like this:
 	#if (TARGETBOARD == NUCLEO_WL55JC1) etc
 with TARGETBOARD etc defined in platform.h

There are further changes for the radio control - both for the GPIO pin allocation and for the
radio control functions. These are undocumented by Seeed and I had to guess.  

- 	radio_board_if_nucleo.h & .c (original provided by ST)
-  	radio_board_if_lora_e5.h & .c (For LoRa-E5)

I added IRQ handlers for all GPIO pins in stm32wlxx_it.c
However I can't get pins which share IRQ handlers to work.

[LoRa-E5_Files.zip](https://github.com/azerimaker/Penguino-STM32WL-LoRa-E5/files/6209071/LoRa-E5_Files.zip)

**Comment by acutetech** [comment]
Hi Orkhan

I have just been told by Seeed that they have published software and some documentation for the LoRa-E5, here:
https://github.com/seeed-lora/LoRa-E5-LoRaWAN-End-Node

Both that page and the radio_board_if.h file are clear that the two RF control pins are PA4 and PA5, not PA4 and PB5 as your contact at Seeed provided in the email below.
The PA5 signal make more sense than PB5 - why would they use PB5 for internal RF control purposes yet also make it available as a pin on the package?

I have tried their code and it seems to work. (I have not figured out why my code also worked when I was using the wrong pin!). But my confidence is not high - for instance they seem to have a bug - see my comment here regarding #define RBI_CONF_RFO
https://forum.seeedstudio.com/t/how-to-program-a-lora-e5/257491/6

I note that the settings of the three RF control signals on generic-node-se are the same as on the Nucleo board. Are you using the same RF design at ST? I cannot find the schematics of their Nucleo board. The LoRa-E5 uses the same values on the first two control signals, but does not have the third.

I would much prefer to have the schematic of the LoRa-E5 - then I would not have to rely on dubious documentation and code!

Your comments welcome.

**Reply by azerimaker** [reply]
Hi @acutetech. 

It's nice to see that Seeed finally came up with an example repo that we can use. Instructions are not bad and well laid out. 

You're right, I'm using the the same pinout as the Nucleo board, for compatibility reasons. The RF path is slightly modified, but software is fully compatible. You can actually download all the design files for [NUCLEO-WL55JC ](https://www.st.com/content/st_com/en/products/evaluation-tools/product-evaluation-tools/mcu-mpu-eval-tools/stm32-mcu-mpu-eval-tools/stm32-nucleo-boards/nucleo-wl55jc.html#cad-resources)under the CAD resources tab on ST's website and compare it with the GN schematics. 

Also asked my contact at Seeed for the Lora-E5 schematics, waiting for his reply. Will follow up on that.
